modules/network/src/main/java/org/apache/ignite/network/internal/netty/NettyClient.java

```diff
@@ -82,12 +85,34 @@ public NettyClient(
         if (clientFuture != null)
             throw new IgniteInternalException("Attempted to start an already started NettyClient");
 
-        clientFuture = NettyUtils.toChannelCompletableFuture(bootstrap.connect(address))
-            .thenApply(ch -> {
-                clientCloseFuture = NettyUtils.toCompletableFuture(ch.closeFuture());
-                channel = ch;
-
-                return new NettySender(channel, serializationRegistry);
+        clientFuture = new CompletableFuture<>();
+
+        NettyUtils.toChannelCompletableFuture(bootstrap.connect(address))
+            .whenComplete((channel, throwable) -> {
+                synchronized (this) {
+                    if (throwable == null) {
+                        CompletableFuture<Void> closeFuture = NettyUtils.toCompletableFuture(channel.closeFuture());
+
+                        if (stopped) {
+                            // Close channel in case if client has been stopped prior to this moment.
+                            channel.close();
+
+                            // Wait for channel to close and then cancel the client future.
+                            closeFuture.whenComplete((unused, ignored) -> {
+                                clientFuture.cancel(true);
```

Comment:
Out of curiosity I read documentation of "cancel" method. Do we really need it? Especially with "true" parameter? What's the point if we're the ones that execute the task? So many questions.

Reply:
We can pass any argument, why not `true`?